[invoke_subgraph] Support unbacked

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149298



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

Differential Revision: [D71420641](https://our.internmc.facebook.com/intern/diff/D71420641)

angelayi committed:
[invoke_subgraph] Support unbacked

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71420641).

angelayi commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -3088,9 +3088,18 @@ def install_subgraph_in_output_graph(
         # installed in the output graph and we can just access the subgraph
         # using the saved attr name.
         from torch._higher_order_ops.utils import has_potential_input_alias_or_mutation
+        from torch._prims_common import clone_preserve_strides
+
+        def _maybe_clone_val(node):
+            example_value = node.meta["example_value"]
+            if isinstance(example_value, torch.Tensor):
+                # We need to clone so that when we retrace, we don't get the
+                # same unbacked symbols
+                return clone_preserve_strides(example_value)
+            return example_value
```
This is a little hard to explain... 

When tracing the function within the invoke_subgraph, we will convert torch._checks into runtime asserts in the graph. This way if we re-run the body_gm graph with fake tensors, we will succeed because the torch._checks will be run as they are in the graph. 

If we retrace body_gm with make_fx, we will run the torch._checks to update the shape env, but this doesn't directly result in a new graph with torch._checks. Instead we need to run the insert_deferred_runtime_asserts pass to insert in torch._checks again based on the `unbacked_bindings` metadata that is added onto nodes.

However, if we retrace the graph with the same inputs that we initially traced with, there's some [caching](https://github.com/pytorch/pytorch/blob/595293316d7e64e32d31716500beae58367409a2/torch/_subclasses/fake_impls.py#L411) going on resulting new unbacked symints not being created, which results in the `unbacked_bindings` metadata not being populated, so then we don't have any torch._checks to add. 

This is what happens currently in the call to [_maybe_fake_tracing](https://github.com/pytorch/pytorch/blob/main/torch/_higher_order_ops/utils.py#L299). The result of the graph from `maybe_fake_tracing` does not contain any of the previous torch._checks. This results in the subsequent [maybe_fake_prop_ignore_unbacked](https://github.com/pytorch/pytorch/blob/main/torch/_higher_order_ops/utils.py#L445-L453) to fail with a DataDependentError.

zou3519 replied:
Can you drop a comment here into the code? I didn't internalize what you wrote but we should have a pointer to it for future reference

angelayi replied:
nvm that broke things

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71420641](https://our.internmc.facebook.com/intern/diff/D71420641)

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71420641](https://our.internmc.facebook.com/intern/diff/D71420641)

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71420641](https://our.internmc.facebook.com/intern/diff/D71420641)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71420641).

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -3088,9 +3088,18 @@ def install_subgraph_in_output_graph(
 
         fake_inputs = [
-            node.meta["example_value"]
+            _maybe_clone_val(node)
```
This is not completely correct, if the inputs have view relationships then that matters if there are mutations

zou3519 replied:
Resolved offline: this is just used for the alias_mutation_check

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71420641](https://our.internmc.facebook.com/intern/diff/D71420641)

[ghstack-poisoned]

zou3519 reviewed (approved):
I suspect we may want to revisit the strategy of when we are running insert_deferred_runtime_asserts in the future, especially for other HOPs. But this LGTM

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71420641](https://our.internmc.facebook.com/intern/diff/D71420641)

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71420641](https://our.internmc.facebook.com/intern/diff/D71420641)

[ghstack-poisoned]

angelayi commented:
@pytorchbot merge

angelayi commented on torch/_higher_order_ops/utils.py:
```diff
@@ -720,11 +724,12 @@ def _tensor_storage(t) -> StorageWeakRef:
-            ]
-            before = [_tensor_version(arg) for arg in cloned]
+            cloned = fake_args
+            # cloned = [
+            #     clone_preserve_strides(arg) if isinstance(arg, torch.Tensor) else arg
+            #     for arg in fake_args
+            # ]
```
oops, i was testing some stuff and forgot to uncomment...

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71420641](https://our.internmc.facebook.com/intern/diff/D71420641)

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71420641](https://our.internmc.facebook.com/intern/diff/D71420641)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71420641).

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71420641](https://our.internmc.facebook.com/intern/diff/D71420641)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71420641).

angelayi committed:
Update on "[invoke_subgraph] Support unbacked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71420641](https://our.internmc.facebook.com/intern/diff/D71420641)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71420641).

angelayi commented:
@pytorchbot merge
